File: app/src/main/java/com/juniperphoton/myersplash/adapter/DownloadsListAdapter.java
package com.juniperphoton.myersplash.adapter;

import android.content.Context;
import android.content.Intent;
import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.facebook.drawee.view.SimpleDraweeView;
import com.juniperphoton.flipperviewlib.FlipperView;
import com.juniperphoton.myersplash.R;
import com.juniperphoton.myersplash.base.App;
import com.juniperphoton.myersplash.model.DownloadItem;
import com.juniperphoton.myersplash.service.BackgroundDownloadService;
import com.juniperphoton.myersplash.utils.DownloadItemTransactionHelper;
import com.juniperphoton.myersplash.utils.Params;
import com.juniperphoton.myersplash.widget.DownloadCompleteView;
import com.juniperphoton.myersplash.widget.DownloadRetryView;
import com.juniperphoton.myersplash.widget.DownloadingView;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

public class DownloadsListAdapter extends RecyclerView.Adapter<DownloadsListAdapter.DownloadItemViewHolder> {
    private Context mContext;
    private List<DownloadItem> mData;

    private DownloadStateChangedCallback mCallback;

    private final static int ITEM = 1;
    private final static int FOOTER = 1 << 1;

    public DownloadsListAdapter(List<DownloadItem> data, Context context) {
        mData = data;
        mContext = context;
    }

    @Override
    public DownloadItemViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        if (viewType == ITEM) {
            View view = LayoutInflater.from(mContext).inflate(R.layout.row_download_item, parent, false);
            int width = mContext.getResources().getDisplayMetrics().widthPixels;
            ViewGroup.LayoutParams params = view.getLayoutParams();
            params.height = (int) (width / 1.7d);
            view.setLayoutParams(params);
            return new DownloadItemViewHolder(view);
        } else if (viewType == FOOTER) {
            View footer = LayoutInflater.from(mContext).inflate(R.layout.row_footer_blank, parent, false);
            return new DownloadItemViewHolder(footer);
        }
        return null;
    }

    @Override
    public void onBindViewHolder(final DownloadItemViewHolder holder, int position) {
        if (getItemViewType(position) == FOOTER) {
            return;
        }
        final DownloadItem item = mData.get(holder.getAdapterPosition());

        holder.downloadCompleteView.setFilePath(item.getFilePath());
        holder.downloadCompleteView.setThemeBackColor(item.getColor());

        holder.draweeView.setImageURI(item.getThumbUrl());
        holder.downloadingView.setProgress(item.getProgress());

        holder.downloadRetryView.setThemeBackColor(item.getColor());
        holder.downloadRetryView.setOnClickDeleteListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                try {
                    mData.remove(holder.getAdapterPosition());
                    notifyItemRemoved(holder.getAdapterPosition());

                    Intent intent = new Intent(App.getInstance(), BackgroundDownloadService.class);
                    intent.putExtra(Params.CANCELED_KEY, true);
                    intent.putExtra(Params.URL_KEY, item.getDownloadUrl());
                    mContext.startService(intent);

                    DownloadItemTransactionHelper.delete(item);
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
        });
        holder.downloadRetryView.setOnClickRetryListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                DownloadItemTransactionHelper.updateStatus(item, DownloadItem.DOWNLOAD_STATUS_DOWNLOADING);
                holder.flipperView.next(item.getStatus());

                Intent intent = new Intent(mContext, BackgroundDownloadService.class);
                intent.putExtra(Params.NAME_KEY, item.getFileName());
                intent.putExtra(Params.URL_KEY, item.getDownloadUrl());
                mContext.startService(intent);
            }
        });

        holder.downloadingView.setProgress(item.getProgress());
        holder.downloadingView.setThemeBackColor(item.getColor());
        holder.downloadingView.setClickCancelListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                DownloadItemTransactionHelper.updateStatus(item, DownloadItem.DOWNLOAD_STATUS_FAILED);
                holder.flipperView.next(item.getStatus());

                Intent intent = new Intent(App.getInstance(), BackgroundDownloadService.class);
                intent.putExtra(Params.CANCELED_KEY, true);
                intent.putExtra(Params.URL_KEY, item.getDownloadUrl());
                mContext.startService(intent);
            }
        });

        int last = item.getLastStatus();
        if (last != item.getStatus()) {
            holder.flipperView.next(item.getStatus());
            item.syncStatus();
        } else {
            holder.flipperView.next(item.getStatus(), false);
        }
    }

    public void setCallback(DownloadStateChangedCallback callback) {
        mCallback = callback;
    }

    private void removeItem(DownloadItem item) {
        for (DownloadItem downloadItem : mData) {
            if (downloadItem.getId().equals(item.getId())) {
                int index = mData.indexOf(downloadItem);
                mData.remove(index);
                break;
            }
        }
        if (mCallback != null) {
            mCallback.onDataChanged();
        }
    }

    @Override
    public int getItemCount() {
        if (mData == null) return 0;
        else return mData.size() + 1;
    }

    @Override
    public int getItemViewType(int position) {
        if (position >= getItemCount() - 1) {
            return FOOTER;
        } else return ITEM;
    }

    public class DownloadItemViewHolder extends RecyclerView.ViewHolder {
        @Nullable
        @BindView(R.id.row_download_item_dv)
        SimpleDraweeView draweeView;

        @Nullable
        @BindView(R.id.row_download_flipper_view)
        FlipperView flipperView;

        @Nullable
        @BindView(R.id.row_downloading_view)
        DownloadingView downloadingView;

        @Nullable
        @BindView(R.id.row_download_retry_view)
        DownloadRetryView downloadRetryView;

        @Nullable
        @BindView(R.id.row_download_complete_view)
        DownloadCompleteView downloadCompleteView;

        public DownloadItemViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }
    }

    public void updateItem(DownloadItem item) {
        int index = mData.indexOf(item);
        if (index >= 0 && index <= mData.size()) {
            Log.d("adapter", "notifyItemChanged:" + index);
            notifyItemChanged(index);
        }
    }

    public void refreshItems(List<DownloadItem> items) {
        mData = items;
        notifyDataSetChanged();
    }

    public void clear() {
        mData.clear();
        notifyDataSetChanged();
    }

    public List<DownloadItem> getData() {
        return mData;
    }

    public interface DownloadStateChangedCallback {
        void onDataChanged();

        void onRetryDownload(String id);
    }
}

File: app/src/main/java/com/juniperphoton/myersplash/adapter/DownloadsListAdapter.kt
package com.juniperphoton.myersplash.adapter

import android.content.Context
import android.content.Intent
import android.support.v7.widget.RecyclerView
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import com.facebook.drawee.view.SimpleDraweeView
import com.juniperphoton.flipperviewlib.FlipperView
import com.juniperphoton.myersplash.App
import com.juniperphoton.myersplash.R
import com.juniperphoton.myersplash.model.DownloadItem
import com.juniperphoton.myersplash.service.BackgroundDownloadService
import com.juniperphoton.myersplash.utils.DownloadItemTransactionUtil
import com.juniperphoton.myersplash.utils.Params
import com.juniperphoton.myersplash.widget.DownloadCompleteView
import com.juniperphoton.myersplash.widget.DownloadRetryView
import com.juniperphoton.myersplash.widget.DownloadingView

import butterknife.BindView
import butterknife.ButterKnife

class DownloadsListAdapter(private var mData: MutableList<DownloadItem>?, private val mContext: Context) :
        RecyclerView.Adapter<DownloadsListAdapter.DownloadItemViewHolder>() {

    private var mCallback: DownloadStateChangedCallback? = null

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): DownloadItemViewHolder? {
        if (viewType == ITEM) {
            val view = LayoutInflater.from(mContext).inflate(R.layout.row_download_item, parent, false)
            val width = mContext.resources.displayMetrics.widthPixels
            val params = view.layoutParams
            params.height = (width / 1.7).toInt()
            view.layoutParams = params
            return DownloadItemViewHolder(view)
        } else if (viewType == FOOTER) {
            val footer = LayoutInflater.from(mContext).inflate(R.layout.row_footer_blank, parent, false)
            return DownloadItemViewHolder(footer)
        }
        return null
    }

    override fun onBindViewHolder(holder: DownloadItemViewHolder, position: Int) {
        if (getItemViewType(position) == FOOTER) {
            return
        }
        val item = mData!![holder.adapterPosition]

        holder.downloadCompleteView?.setFilePath(item.filePath?:"")
        holder.downloadCompleteView?.setThemeBackColor(item.color)

        holder.draweeView?.setImageURI(item.thumbUrl)
        holder.downloadingView?.setProgress(item.progress)

        holder.downloadRetryView?.setThemeBackColor(item.color)
        holder.downloadRetryView?.setOnClickDeleteListener(View.OnClickListener {
            try {
                mData!!.removeAt(holder.adapterPosition)
                notifyItemRemoved(holder.adapterPosition)

                val intent = Intent(App.instance, BackgroundDownloadService::class.java)
                intent.putExtra(Params.CANCELED_KEY, true)
                intent.putExtra(Params.URL_KEY, item.downloadUrl)
                mContext.startService(intent)

                DownloadItemTransactionUtil.delete(item)
            } catch (e: Exception) {
                e.printStackTrace()
            }
        })
        holder.downloadRetryView?.setOnClickRetryListener(View.OnClickListener {
            DownloadItemTransactionUtil.updateStatus(item, DownloadItem.DOWNLOAD_STATUS_DOWNLOADING)
            holder.flipperView?.next(item.status)

            val intent = Intent(mContext, BackgroundDownloadService::class.java)
            intent.putExtra(Params.NAME_KEY, item.fileName)
            intent.putExtra(Params.URL_KEY, item.downloadUrl)
            mContext.startService(intent)
        })

        holder.downloadingView?.setProgress(item.progress)
        holder.downloadingView?.setThemeBackColor(item.color)
        holder.downloadingView?.setClickCancelListener(View.OnClickListener {
            DownloadItemTransactionUtil.updateStatus(item, DownloadItem.DOWNLOAD_STATUS_FAILED)
            holder.flipperView?.next(item.status)

            val intent = Intent(App.instance, BackgroundDownloadService::class.java)
            intent.putExtra(Params.CANCELED_KEY, true)
            intent.putExtra(Params.URL_KEY, item.downloadUrl)
            mContext.startService(intent)
        })

        val last = item.lastStatus
        if (last != item.status && last != DownloadItem.DISPLAY_STATUS_NOT_SPECIFIED) {
            holder.flipperView?.next(item.status)
            item.syncStatus()
        } else {
            holder.flipperView?.next(item.status, false)
            item.syncStatus()
        }
    }

    fun setCallback(callback: DownloadStateChangedCallback) {
        mCallback = callback
    }

    private fun removeItem(item: DownloadItem) {
        for (downloadItem in mData!!) {
            if (downloadItem.id == item.id) {
                val index = mData!!.indexOf(downloadItem)
                mData!!.removeAt(index)
                break
            }
        }
        if (mCallback != null) {
            mCallback!!.onDataChanged()
        }
    }

    override fun getItemCount(): Int {
        if (mData == null)
            return 0
        else
            return mData!!.size + 1
    }

    override fun getItemViewType(position: Int): Int {
        if (position >= itemCount - 1) {
            return FOOTER
        } else
            return ITEM
    }

    inner class DownloadItemViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        @BindView(R.id.row_download_item_dv)
        @JvmField var draweeView: SimpleDraweeView? = null

        @BindView(R.id.row_download_flipper_view)
        @JvmField var flipperView: FlipperView? = null

        @BindView(R.id.row_downloading_view)
        @JvmField var downloadingView: DownloadingView? = null

        @BindView(R.id.row_download_retry_view)
        @JvmField var downloadRetryView: DownloadRetryView? = null

        @BindView(R.id.row_download_complete_view)
        @JvmField var downloadCompleteView: DownloadCompleteView? = null

        init {
            ButterKnife.bind(this, itemView)
        }
    }

    fun updateItem(item: DownloadItem) {
        val index = mData!!.indexOf(item)
        if (index >= 0 && index <= mData!!.size) {
            Log.d("adapter", "notifyItemChanged:" + index)
            notifyItemChanged(index)
        }
    }

    fun refreshItems(items: MutableList<DownloadItem>) {
        mData = items
        notifyDataSetChanged()
    }

    fun clear() {
        mData!!.clear()
        notifyDataSetChanged()
    }

    val data: List<DownloadItem>?
        get() = mData

    interface DownloadStateChangedCallback {
        fun onDataChanged()

        fun onRetryDownload(id: String)
    }

    companion object {
        private val ITEM = 1
        private val FOOTER = 1 shl 1
    }
}
